### PR TITLE
configure machine deployment

### PR DESCRIPTION
**Pull request by amy**
Configures machine deployment YAML. 
Addresses: https://github.com/kubernetes-sigs/cluster-api-provider-docker/issues/29

### CODE REVIEW

File: cmd/capdctl/main.go [hunk 1]
```diff
@@ -47,6 +47,21 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
...
+type machineDeyploymentOptions struct {
+	name, namespace, clusterName, controlPlaneVersion, kubeletVersion *string
+	replicas                                                          *int32
+}
+
+func (mo *machineDeyploymentOptions) initFlags(fs *flag.FlagSet) {
+	mo.name = fs.String("name", "my-machine", "The name of the machine")
+	mo.namespace = fs.String("namespace", "my-namespace", "The namespece of the machine")
+	mo.clusterName = fs.String("cluster-name", "my-cluster", "The name of the cluster the machine belongs to")
+	mo.controlPlaneVersion = fs.String("control-plane-version", "v1.14.2", "The Kubernetes control plane version to run")
```
Comment (chuckha) [inline cmd/capdctl/main.go]: MachineDeployments will only (for right now) manage worker nodes, we don't need to configure the control plane version on them

File: cmd/capdctl/main.go [hunk 2]
```diff
@@ -47,6 +47,21 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
 	mo.version = fs.String("version", "v1.14.2", "The Kubernetes version to run")
 }
 
+type machineDeyploymentOptions struct {
+	name, namespace, clusterName, controlPlaneVersion, kubeletVersion *string
+	replicas                                                          *int32
+}
+
+func (mo *machineDeyploymentOptions) initFlags(fs *flag.FlagSet) {
+	mo.name = fs.String("name", "my-machine", "The name of the machine")
```
Comment (chuckha) [inline cmd/capdctl/main.go]: This is the name of the machine deployment, not the name of the machine

File: cmd/capdctl/main.go [hunk 3]
```diff
@@ -47,6 +47,21 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
...
+	mo.namespace = fs.String("namespace", "my-namespace", "The namespece of the machine")
```
Comment (chuckha) [inline cmd/capdctl/main.go]: s/machine/machine deployment/

File: cmd/capdctl/main.go [hunk 4]
```diff
@@ -47,6 +47,21 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
...
+	mo.clusterName = fs.String("cluster-name", "my-cluster", "The name of the cluster the machine belongs to")
```
Comment (chuckha) [inline cmd/capdctl/main.go]: same, machine should be machine deployment

File: cmd/capdctl/main.go [hunk 5]
```diff
@@ -47,6 +47,21 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
...
+	mo.controlPlaneVersion = fs.String("control-plane-version", "v1.14.2", "The Kubernetes control plane version to run")
+	mo.kubeletVersion = fs.String("kubelet-version", "v1.14.2", "The Kubernetes kubelet version to run")
+	replicas := int32(*fs.Int("replicas", 1, "The number of replicas"))
```
Comment (chuckha) [inline cmd/capdctl/main.go]: god i love go so much.

I would probably recommend passing in an int here for readability and converting it inside the machine deployment function.

but it doesn't really matter, that's just a minor style preference

File: cmd/capdctl/main.go [hunk 6]
```diff
@@ -143,6 +158,41 @@ spec:
...
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      *opts.name,
+			Namespace: *opts.namespace,
+			Labels: map[string]string{
+				"cluster.k8s.io/cluster-name": *opts.clusterName,
+				"set":                         "node",
+			},
+		},
+		Spec: v1alpha1.MachineDeploymentSpec{
```
Comment (chuckha) [inline cmd/capdctl/main.go]: The MachineDeploymentSpec is missing the

```
selector:
    matchLabels:
...
```
section.

File: cmd/capdctl/main.go [hunk 7]
```diff
@@ -143,6 +158,41 @@ spec:
...
+			Replicas: opts.replicas,
+			Template: v1alpha1.MachineTemplateSpec{
```
Comment (chuckha) [inline cmd/capdctl/main.go]: This template is missing the Metadata object

File: cmd/capdctl/main.go [hunk 8]
```diff
@@ -143,6 +158,41 @@ spec:
...
+				Spec: v1alpha1.MachineSpec{
```
Comment (chuckha) [inline cmd/capdctl/main.go]: I lied a little bit, the provider spec needs to be a nil object

```
ProviderSpec: v1alpha1.ProviderSpec{},
```

File: cmd/capdctl/main.go [hunk 9]
```diff
@@ -143,6 +158,41 @@ spec:
...
+					Versions: v1alpha1.MachineVersionInfo{
+						ControlPlane: *opts.controlPlaneVersion,
```
Comment (chuckha) [inline cmd/capdctl/main.go]: we don't ever want to set the control plane version since this machine deployment will only manage worker nodes

**Comment by neolit123** [comment]
/test pull-cluster-api-provider-docker-verify
testing if the CI works.

**Comment by neolit123** [comment]
```
Found trailing whitespace in the follow files:
./README.md:9:- `go,  1.12+` 
./README.md:47:1. Install capdctl: 
Verifying new lines at end of files...
Found whitespace errors!
```
it works and it caught errors from previous PRs that already merged.

File: cmd/capdctl/main.go [hunk 10]
```diff
@@ -47,6 +47,19 @@ func (mo *machineOptions) initFlags(fs *flag.FlagSet) {
...
 
+type machineDeyploymentOptions struct {
+	name, namespace, clusterName, kubeletVersion *string
+	replicas                                     *int
+}
+
+func (mo *machineDeyploymentOptions) initFlags(fs *flag.FlagSet) {
+	mo.name = fs.String("name", "my-machine-deployment", "The name of the machine deployment")
+	mo.namespace = fs.String("namespace", "my-namespace", "The namespece of the machine deployment")
+	mo.clusterName = fs.String("cluster-name", "my-cluster", "The name of the cluster the machine belongs to")
```
Comment (chuckha) [inline cmd/capdctl/main.go]: this should be either

> the name of the cluster the machines will belong to 

or

> the name of the cluster the machine deployment creates machines for

File: cmd/capdctl/main.go [hunk 11]
```diff
@@ -124,6 +144,9 @@ subcommands are:
 
   cluster - Write a capd cluster object to stdout
     example: capdctl cluster -cluster-name my-cluster -namespace my-namespace | kubectl apply -f -
+  
+  machine-deployment - Write a machine deployment objecto to stdout
```
Comment (chuckha) [inline cmd/capdctl/main.go]:
```suggestion
  machine-deployment - Write a machine deployment object to to stdout
```

File: cmd/capdctl/main.go [hunk 12]
```diff
@@ -124,6 +144,9 @@ subcommands are:
...
+    example: capdctl machine-deployment -name my-machine-deployment -cluster-name my-cluster -namespace my-namespace -kubelet-version v1.14.2 -replicas 1| kubectl apply -f -
```
Comment (chuckha) [inline cmd/capdctl/main.go]:
```suggestion
    example: capdctl machine-deployment -name my-machine-deployment -cluster-name my-cluster -namespace my-namespace -kubelet-version v1.14.2 -replicas 1 | kubectl apply -f -
```

**Comment by chuckha** [comment]
I don't usually do this but the build is broken due to CI and this has the fixes. The other exception is that this is still a new project and we're looking for reviewers!

/lgtm

**Pull request merged**